### PR TITLE
internal/envoy: fix TCPProxy weights bug

### PR DESCRIPTION
**Pull request by skriss**
Fixes a bug where when a TCPProxy had more than one
service defined, any service with a zero weight was
updated to have a weight of 1. The correct behavior
is that if all service have no weights, then they
should be evenly weighted, but if some have non-zero
weights and others have zero weights, they stay as-is.

Signed-off-by: Steve Kriss <krisss@vmware.com>

Notes for reviewers:
- The desired behavior that I describe in the commit message is not explicitly documented anywhere for TCPProxy, however it is for HTTPProxy (and my assumption is that the behavior should be the same between them): https://projectcontour.io/docs/v1.18.0/config/request-routing/#upstream-weighting
- This came up while working on supporting weighted forward to's for TLSRoute (#3440).

### CODE REVIEW

**Review by stevesloka — APPROVED** [review]

File: internal/envoy/v3/listener.go [hunk 1]
```diff
@@ -460,52 +460,72 @@ func TCPProxy(statPrefix string, proxy *dag.TCPProxy, accesslogger []*accesslog.
...
+	// If some clusters have non-zero weights, then drop
+	// any others that have zero weights since they shouldn't
+	// get traffic (note that TCPProxy weighted clusters can't
+	// have zero weights, unlike HTTP route weighted clusters,
+	// so we can't include them with a zero weight). Also note
+	// that if all clusters have zero weights, then we keep them
+	// all and evenly weight them below.
+	var keepClusters []*dag.Cluster
+	for _, c := range proxy.Clusters {
+		if totalWeight > 0 && c.Weight == 0 {
```
Comment (sunjayBhatia) [inline internal/envoy/v3/listener.go:484]: just a thought how to do this in one pass (havent tested it tho)

```
var totalWeight uint32
var weightedClusters []*dag.Cluster
for _, c := range proxy.Clusters {
  if c.Weight != 0 {
    weightedClusters = append(weightedClusters, c)
  }
  totalWeight += c.Weight
}

var actualClusters []*dag.Cluster
if totalWeight == 0 {
  actualClusters = proxy.Clusters
} else {
  actualClusters = weightedClusters
}

switch len(actualClusters) {
...
```

could rename `weightedClusters` to `clustersWithWeight` too

**Review by sunjayBhatia — APPROVED** [review]
just one thought about how to maybe do the processing with one loop, otherwise lgtm! nice fix

File: internal/envoy/v3/listener.go (hunk 1 above)
Comment (skriss) [reply internal/envoy/v3/listener.go]: :+1: I prioritized readability over perf here, but made an update here along the lines of ^^^ that I think is still easy to read/follow.

Comment (sunjayBhatia) [reply internal/envoy/v3/listener.go]: nice! yeah good adaptation

**Pull request merged**